### ISSUE TITLE
TrackingMonitor scales badly with the number of streams/threads

### ISSUE THREAD

**Issue by makortel** [cms-sw/cmssw#32953]
I accidentally noticed that `TrackingMonitor` scales badly with the number of streams/threads in a Run 3 trackingOnly workflow (11634.1 step 3 in 11_2_0) that was using ttbar+PU50 events as an input. Below are per-event times for processing 64 events with 1, 2, 4, 8, and 16 threads
```
1  TimeReport   0.701662  TrackerCollisionSelectedTrackMonCommongeneralTracks
2  TimeReport   0.781637  TrackerCollisionSelectedTrackMonCommongeneralTracks
4  TimeReport   1.291316  TrackerCollisionSelectedTrackMonCommongeneralTracks
8  TimeReport   4.178550  TrackerCollisionSelectedTrackMonCommongeneralTracks
16 TimeReport  67.869327  TrackerCollisionSelectedTrackMonCommongeneralTracks
```
(the first column is the number of threads, and the third column the time per event in seconds). In this measurement the machine was otherwise empty, so the CPU frequency scaling impacts the result but I'd claim effect is too big to be explained by that.

I recorded IgProf profiles of 1 thread and 16 thread cases
* 1 thread https://mkortela.web.cern.ch/mkortela/cgi-bin/navigator/1120/reco_trackingOnly_perf/30
* 16 treads https://mkortela.web.cern.ch/mkortela/cgi-bin/navigator/1120/reco_trackingOnly_perf_nth16/16

Essentially all the cost comes from the `tbb::spin_mutex` in `MonitorElement::getBinContent()`
https://mkortela.web.cern.ch/mkortela/cgi-bin/navigator/1120/reco_trackingOnly_perf/47
https://mkortela.web.cern.ch/mkortela/cgi-bin/navigator/1120/reco_trackingOnly_perf_nth16/19

The cause for the lock contention appears to be this double-loop
https://github.com/cms-sw/cmssw/blob/ba6e8604a35283e39e89bc031766843d0afc3240/DQM/TrackingMonitor/src/TrackAnalyzer.cc#L2037-L2059

Thinking out loud some options
- Does the double-loop need to be done for each event?
- If nothing else works, the module should be made to `edm::one` (because the assumption of "low contention of the `tbb::spin_mutex`" does not hold for this module)

**Comment by makortel** [comment]
assign dqm

**Comment by Dr15Jones** [comment]
The loop pointed out by Matti happens each time just 1 'eta/phi' value is added which is not needed. Instead, there are only 3 bins which have to be checked. One for each of the histograms: tkmes.TrackEtaPhi, tkmes.TrackEtaPhiInverted and tkmes.TrackEtaPhiInvertedoutofphase. It is even possible that the same bin would be filled. Therefore one could

- setup a std::set to hold the 'universal bin' ID
- call `MonitorElement::getBin` using the same value as passed to `Fill` for each of the three histograms and adding that to the std::set
- loop over the values in the set and call `getBinContent` once and only once for each histogram. Then call `setBinContent` for that bin on the histograms tkmes.TkEtaPhi_RelativeDifference_byFoldingmap and tkmes.TkEtaPhi_RelativeDifference_byFoldingmap_op

To work all bins of tkmes.TkEtaPhi_RelativeDifference_byFoldingmap and tkmes.TkEtaPhi_RelativeDifference_byFoldingmap_op have to be initialized to 1 right after booking.

**Comment by davidlange6** [comment]
Or just do that double loop at end of job rather than once per track?

**Comment by jfernan2** [comment]
@mtosi  could you please provide some feedback?

**Comment by mtosi** [comment]
w/ @sroychow, @mmusich and alessandro we think we have identified this "bad guy"
to be the source of both the multithread issue and the CPU timing one as well

we are fixing it, and a new PR should come "shortly"

**Comment by mmusich** [comment]
This is tackled at  #33430.
Experts can please have a look.

**Comment by makortel** [comment]
> This is tackled at #33430.
> Experts can please have a look.

Thanks!

**Comment by makortel** [comment]
https://github.com/cms-sw/cmssw/pull/33430 is merged, we can close the issue

**Comment by jfernan2** [comment]
+1
Thanks for the prompt action!!
